chalk-ir/src/visit/boring_impls.rs

```diff
@@ -0,0 +1,341 @@
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        let interner = visitor.interner();
+        let mut result = R::new();
+
+        for p in self.iter(interner) {
+            result = result.and_then(|| p.visit_with(visitor, outer_binder));
+        }
+
+        result
+    }
+}
+
+impl<I: Interner> Visit<I> for Goals<I> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        let interner = visitor.interner();
+        let mut result = R::new();
+
+        for p in self.iter(interner) {
+            result = result.and_then(|| p.visit_with(visitor, outer_binder));
+        }
+
+        result
+    }
+}
+
+#[macro_export]
+macro_rules! const_visit {
+    ($t:ty) => {
+        impl<I: Interner> $crate::visit::Visit<I> for $t {
+            fn visit_with<'i, R: VisitResult>(
+                &self,
+                _visitor: &mut dyn ($crate::visit::Visitor<'i, I, Result = R>),
+                _outer_binder: DebruijnIndex,
+            ) -> R
+            where
+                I: 'i,
+            {
+                R::new()
+            }
+        }
+    };
+}
+
+const_visit!(UniverseIndex);
+const_visit!(usize);
+const_visit!(PlaceholderIndex);
+const_visit!(QuantifierKind);
+const_visit!(DebruijnIndex);
+const_visit!(chalk_engine::TableIndex);
+const_visit!(chalk_engine::TimeStamp);
+const_visit!(());
+
+#[macro_export]
+macro_rules! id_visit {
+    ($t:ident) => {
+        impl<I: Interner> $crate::visit::Visit<I> for $t<I> {
+            fn visit_with<'i, R: VisitResult>(
+                &self,
+                _visitor: &mut dyn ($crate::visit::Visitor<'i, I, Result = R>),
+                _outer_binder: DebruijnIndex,
+            ) -> R
+            where
+                I: 'i,
+            {
+                R::new()
+            }
+        }
+    };
+}
+
+id_visit!(ImplId);
+id_visit!(StructId);
+id_visit!(TraitId);
+id_visit!(AssocTypeId);
+
+impl<I: Interner> SuperVisit<I> for ProgramClause<I> {
+    fn super_visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        match self {
+            ProgramClause::Implies(pci) => pci.visit_with(visitor, outer_binder),
+            ProgramClause::ForAll(pci) => pci.visit_with(visitor, outer_binder),
+        }
+    }
+}
+
+impl<I: Interner> Visit<I> for PhantomData<I> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        _visitor: &mut dyn Visitor<'i, I, Result = R>,
+        _outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        R::new()
+    }
+}
+
+impl<I: Interner, T, L> Visit<I> for ParameterKind<T, L>
```

Comment:
can this not be derived..?

Reply:
Derive is only for types that have one type parameter that either has an interner or is the interner. I don't know why is it so strict, but I guess there was a reason for that?

Reply:
Oh. Well, it's almost always sufficient I guess -- and the more general pattern isn't as obvious -- but I think we could extend it readily enough. e.g., to any number of type parameters that are all `HasInterner`, which would apply here, right? But if it's only the one impl, maybe not worth it (I imagine this is why we didn't bother the first time)